src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -580,18 +587,19 @@ private void AddHeaderInfo(HeaderDescriptor descriptor, HeaderStoreItemInfo sour
         {
             HeaderStoreItemInfo destinationInfo = CreateAndAddHeaderToStore(descriptor);
 
-            // We have custom header values. The parsed values are strings.
+            // Always copy raw values
+            destinationInfo.RawValue = CloneStringHeaderInfoValues(sourceInfo.RawValue);
+
             if (descriptor.Parser == null)
             {
-                Debug.Assert((sourceInfo.RawValue == null) && (sourceInfo.InvalidValue == null),
-                    "No raw or invalid values expected for custom headers.");
-
+                // We have custom header values. The parsed values are strings.
```

Comment:
Why did this assert change?

Reply:
Because it was wrong :)

It turns out we only ever hit AddHeaders code path for two things:
1) Adding in the DefaultRequestHeaders
2) Forwarding headers for a decompressed response in SocketsHttpHandler

For (2), the headers are all raw response headers, so in AddHeaders above, the `if (sourceValue is HeaderStoreItemInfo info)` check will always be false and we'll never hit this AddHeaderInfo path.

For (1), the headers are all forced to be in HeaderStoreItemInfo instances (something else to revisit), and we were only getting here after calling ParseRawHeaderValues, which for a custom header with no validation would basically transfer the raw value to the parsed field.

But I removed that ParseRawHeaderValues call, so now we may still have raw values here.

Reply:
Yeah, makes sense, thanks.